nsdb-core/src/main/scala/io/radicalbit/nsdb/protocol/MessageProtocol.scala

```diff
@@ -32,102 +33,124 @@ object MessageProtocol {
+    case class DeleteAllSchemas(db: String, namespace: String)                                extends NSDbSerializable
 
     case class GetCountWithLocations(db: String, namespace: String, metric: String, locations: Seq[Location])
-    case class AddRecordToShard(db: String, namespace: String, location: Location, bit: Bit)
+        extends NSDbSerializable
+    case class AddRecordToShard(db: String, namespace: String, location: Location, bit: Bit) extends NSDbSerializable
     case class DeleteRecordFromShard(db: String, namespace: String, shardKey: Location, bit: Bit)
-    case class DeleteAllMetrics(db: String, namespace: String)
+        extends NSDbSerializable
+    case class DeleteAllMetrics(db: String, namespace: String) extends NSDbSerializable
 
-    case object GetNodeChildActors
+    case object GetNodeChildActors extends NSDbSerializable
     case class NodeChildActorsGot(metadataCoordinator: ActorRef,
                                   writeCoordinator: ActorRef,
                                   readCoordinator: ActorRef,
                                   publisher: ActorRef)
+        extends NSDbSerializable
 
-    case class SubscribeMetricsDataActor(actor: ActorRef, nodeName: String)
-    case class SubscribeCommitLogCoordinator(actor: ActorRef, nodeName: String)
-    case class SubscribePublisher(actor: ActorRef, nodeName: String)
+    case class SubscribeMetricsDataActor(actor: ActorRef, nodeName: String)     extends NSDbSerializable
+    case class SubscribeCommitLogCoordinator(actor: ActorRef, nodeName: String) extends NSDbSerializable
+    case class SubscribePublisher(actor: ActorRef, nodeName: String)            extends NSDbSerializable
 
-    case object GetConnectedDataNodes
+    case object GetConnectedDataNodes extends NSDbSerializable
 
-    case object GetMetricsDataActors
-    case object GetCommitLogCoordinators
-    case object GetPublishers
+    case object GetMetricsDataActors     extends NSDbSerializable
+    case object GetCommitLogCoordinators extends NSDbSerializable
+    case object GetPublishers            extends NSDbSerializable
 
-    case class Migrate(inputPath: String)
+    case class Migrate(inputPath: String) extends NSDbSerializable
   }
 
   /**
     * events received from nsdb actors.
     */
   object Events {
 
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+    @JsonSubTypes(
+      Array(
+        new JsonSubTypes.Type(value = classOf[MetricNotFound], name = "MetricNotFound"),
+        new JsonSubTypes.Type(value = classOf[Generic], name = "Generic")
+      ))
     sealed trait ErrorCode
-    case class MetricNotFound(metric: String) extends ErrorCode
-    case object Generic                       extends ErrorCode
+    case class MetricNotFound(metric: String) extends ErrorCode with NSDbSerializable
+    case class Generic()                      extends ErrorCode with NSDbSerializable
```

Comment:
why case class andnot case object?

Reply:
Good point. It would be tricky (do not know if possible actually) to write a custom serializer for an ADT with mixed case class and case object as children. 
So either all case class or all case object children are allowed but not mixed.

I don't like it, and maybe technical debts issue  can be elaborated and fixed